Pythonise the documentation according to #4187: Basic Tutorial (cython_tutorial.rst) (#4226)
cython/cython: docs/src/tutorial/cython_tutorial.rst

Here's a small example showing some of what can be done. It's a routine for
finding prime numbers. You tell it how many primes you want, and it returns
them as a Python list.
 
.. tabs::
    .. group-tab:: Cython

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
            :linenos:
            :caption: primes.pyx

    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py

>             :linenos:

Reviewer: Couldn't you use linenos to not show the `import cython` that would make tests pass?

Reviewer 2: I think we decided to modify the test-runner to auto-add `import cython`. Although that isn't included in this PR (so currently the tests will probably fail)

Author: > Couldn't you use linenos to not show the `import cython` that would make tests pass?
> 
> Edit: or `linenos-start`

Yes, we can, and I have already said this in the issue. The fact is that then all the other lines will be displaced, and it will be very inconvenient to edit them.
Look [here](https://github.com/cython/cython/issues/4187#issuecomment-859973066)

            :caption: primes.py

You'll see that it starts out just like a normal Python function definition,
except that the parameter ``nb_primes`` is declared to be of type ``int``. This
means that the object passed will be converted to a C integer (or a
``TypeError.`` will be raised if it can't be).

Now, let's dig into the core of the function:

.. tabs::
    .. group-tab:: Cython

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
            :lines: 2,3
            :dedent:
            :lineno-start: 2

        Lines 2 and 3 use the ``cdef`` statement to define some local C variables.
        The result is stored in the C array ``p`` during processing,
        and will be copied into a Python list at the end (line 26).

    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
            :lines: 2,3
            :dedent:
            :lineno-start: 2

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
            :lines: 11,12
            :dedent:
            :lineno-start: 11
        
        Lines 2, 3, 11 and 12 use the variable annotations

>         to definesome local C variables.

Reviewer 2: ```suggestion
        to define some local C variables.
```

Also, the text that changes with the code tab is a good idea, but I wonder if it should be present as "part of" the tab.

Author: If you downloaded this PR to yourself, and watched how it looks, then it would become obvious to you that there is no need to draw a border around changing content. On the other hand, this is a bit of a matter of taste, but it seems to me that this is superfluous. When changed between versions, the user will see everything for himself.

Author: suggestion [here](https://github.com/cython/cython/pull/4226/commits/1eff82edd97ed23df73088b6caa3a70f95e64d26)

        The result is stored in the C array ``p`` during processing,
        and will be copied into a Python list at the end (line 26).

.. NOTE:: You cannot create very large arrays in this manner, because
          they are allocated on the C function call :term:`stack<Stack allocation>`,
          which is a rather precious and scarce resource.
          To request larger arrays,
          or even arrays with a length only known at runtime,
          you can learn how to make efficient use of
          :ref:`C memory allocation <memory_allocation>`,
          :ref:`Python arrays <array-array>`
          or :ref:`NumPy arrays <memoryviews>` with Cython.

.. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
    :lines: 5,6
    :dedent:
    :lineno-start: 5

As in C, declaring a static array requires knowing the size at compile time.
We make sure the user doesn't set a value above 1000 (or we would have a
segmentation fault, just like in C)

.. tabs::
    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
            :lines: 8,9
            :dedent:
            :lineno-start: 8

>         If we run the code from python, then we must initialize the array,
>         the easiest way is to fill it with zeros as on lines 8-9.
>         If we compile this, then the array will already be filled with
>         random data from the allocated memory, since we will change them anyway,
>         we do not need to overwrite this random data.

Reviewer 3: ```suggestion
        When we run this from Python we have to initialize the list,
        which is easiest done filling it with 0s (as seen on lines 8-9).
        When we compile this with Cython on the other hand, the array
        will be a 1000 items long with random data from previous allocations
        that we will overwrite.
```

Reviewer 4: > When we compile this with Cython on the other hand, the array
> will be a 1000 items long with random data from previous allocations
> that we will overwrite.

random -> arbitrary. The data is not random, that's a common misperception that we should not spread here.

```suggestion
        When we run this code from Python, we have to initialize the array.
        This is most easily done by filling it with zeros (as seen on line 8-9).
        When we compile this with Cython, on the other hand, the array will
        behave as in C.  It is allocated on the function call stack with a fixed
        length of 1000 items that contain arbitrary data from the last time that
        memory was used.  We will then overwrite those items in our calculation.
```

Author: > The data is not random, that's a common misperception that we should not spread here.

or we can say pseudo-random to be true.

Reviewer 3: No Scoder is right in this, it appears random. But it is not *randomly generated*, it is random in that it has no meaning -> arbitrary.

Author: > it is not _randomly generated_

Of course, I did not say that this is the result of the generation of pre-random numbers

> it has no meaning

Not completely true, this data made sense until it was deallocated, and we can even try to read it and get some meaningful information.

            :lineno-start: 8

        If we run the code from python, then we must initialize the array,
        the easiest way is to fill it with zeros as on lines 8-9.
        If we compile this, then the array will already be filled with
        random data from the allocated memory, since we will change them anyway,

>         we do not need to overwrite this random data.

Reviewer 2: I found this bit slightly hard to read, because the initial content is quite different to the non-pure case. I've made some changes to the code that'd hopefully get rid of it.

Author: Not quite sure what you changed? @da-woods you can suggest a different, clearer version of the text.

Author: Original modified, I think I can close this
look [here](https://github.com/cython/cython/pull/4226#discussion_r651328644)

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
            :lines: 10-13
            :dedent:
            :lineno-start: 10

> Lines 11-13 set up for a loop which will test candidate numbers for primeness
> until the required number of primes has been found.

Reviewer 3: It is rather confusing to call it a for loop.
```suggestion
Lines 11-13 sets up a while loop that starts looping through all numbers looking for primes,
stopping when the required number of primes have been found.
```

Author: "Loop-looping" doesn't sound very good.

Ok, [here](https://github.com/cython/cython/pull/4226/commits/5a44abb98f64e723283c6460d119a90979a8c899) is some mix, I think it worked out quite well.

.. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
    :lines: 14-17
    :dedent:
    :lineno-start: 14

> Lines 15-16, which try dividing a candidate by all the primes found so far,

Reviewer 3: ```suggestion
Lines 15-16, which try to divide a candidate by all the primes found so far,
```

See, it works! And if you're curious about how much work Cython has saved you,
take a look at the C code generated for this module.

Cython has a way to visualise where interaction with Python objects and
Python's C-API is taking place. For this, pass the
``annotate=True`` parameter to ``cythonize()``. It produces a HTML file. Let's see:

.. tabs::
    .. group-tab:: Cython

        .. figure:: htmlreport.png
            :scale: 90 %

    .. group-tab:: Pure Python

        .. figure:: htmlreport_pure.png

>             :scale: 90 %

Reviewer 2: Image (for pure) shows filenames from earlier revision

Author: Although this almost does not affect anything, I will redo it now.

Author: [here](https://github.com/cython/cython/pull/4226/commits/0614033ac1d1f776a2d4d5e698e12e42fc66ee88)

> .. tabs::
>     .. group-tab:: Cython
> 
>         .. figure:: htmlreport.png
>             :scale: 90 %
> 
>     .. group-tab:: Pure Python
> 
>         .. figure:: htmlreport_pure.png
>             :scale: 90 %

Reviewer 4: I really don't like the `_pure` suffix. We should avoid it if possible and use balanced names instead (where they make sense).
```suggestion
.. tabs::
    .. group-tab:: Cython

        .. figure:: htmlreport_pyx.png
            :scale: 90 %

    .. group-tab:: Pure Python

        .. figure:: htmlreport_py.png
            :scale: 90 %
```

Author: I accept the suggestion, at the same time changing [this](https://github.com/cython/cython/pull/4226/commits/b862b36044d95abdfbad2f4b95efc1d833f0dcd4) and [that](https://github.com/cython/cython/pull/4226/commits/5febe377a28cb574ce9b60b96d7331388cfc685d)

it involves the creation of a Python object. But the line ``if n % i == 0:``, why?
We can examine the generated C code to understand:

> .. tabs::
>     .. group-tab:: Cython
> 
>         .. figure:: python_division.png
> 
>     .. group-tab:: Pure Python
> 
>         .. figure:: python_division_pure.png

Reviewer 4: ```suggestion
.. tabs::
    .. group-tab:: Cython

        .. figure:: python_division_pyx.png

    .. group-tab:: Pure Python

        .. figure:: python_division_py.png
```

Author: accepted, at the same time changing [this](https://github.com/cython/cython/pull/4226/commits/cc3ab8163ff55dd1d8ac0e8e76063a76b0f24f9c) and [that](https://github.com/cython/cython/pull/4226/commits/c3cda8b03606ae58023f1134f3be221658b2076e)

We can see that some checks happen. Because Cython defaults to the
Python behavior, the language will perform division checks at runtime,
just like Python does. You can deactivate those checks by using the
:ref:`compiler directives<compiler-directives>`.

Now let's see if we get a speed increase even if there is a division check.
Let's write the same program, but in Python:

.. literalinclude:: ../../examples/tutorial/cython_tutorial/primes_python.py
    :caption: primes_python.py / primes_python_compiled.py

> It is possible to take a plain ``.py`` file and to compile it with Cython.

Reviewer 2: "plain .py file" -> "unannonated .py file" or "untyped .py file" (since the "pure" case kind of is a plain py file anyway)

Author: [here](https://github.com/cython/cython/pull/4226/commits/943bc7031b97266dc0eb3817a180628dd8cec5ca)

> Let's create a copy of ``primes_python`` and name it ``primes_python_compiled`` and
> compile it with Cython (without changing the code).

Reviewer 4: ```suggestion
Let's create a copy of ``primes_python`` and name it ``primes_python_compiled``
to be able to compare it to the (non-compiled) Python module.
Then we compile that file with Cython, without changing the code.
```

Now the ``setup.py`` looks like this:

.. tabs::
    .. group-tab:: Cython

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/setup_primes.py

    .. group-tab:: Pure Python

>         .. literalinclude:: ../../examples/tutorial/cython_tutorial/setup_primes_pure.py

Reviewer 4: I'm not sure if it's worth putting these in external files. The main point here are the file names that get passed into `cythonize()`, and compiling the file won't validate them. It's easier to see that they are correct when inlined here.

Author: @scoder it's just not very convenient to change and maintain them if they are in a block of tabs. It is possible to remove .py from the name so that the tests do not occur.

Author: And look at [this file](https://github.com/cython/cython/blob/master/docs/examples/tutorial/cython_tutorial/setup.py) then, it was there before and everything was fine

Reviewer 4: It just seems wrong to me to have these under a different name than `setup.py` and to have them in files at all, where they expose a higher risk of not matching the text. It's too easy to rename one of the other files and forget to adapt this one.

The fact that this has been done before does not mean that it was previously a good decision.

Author: > It's too easy to rename one of the other files and forget to adapt this one.

I cannot disagree. Putting this in separate files allows you to test them, but it probably makes very little sense, even more so compared to the possibility of forgetting to update them.
In general, I think I will agree, and in the near future I will add them as text.

Reviewer 4: Yes, I think it would be better to have the setup.py scripts included as text, not files.

Author: resolved [here](https://github.com/cython/cython/pull/4226/commits/e1cb02468f09921fb8f019a1d0efc3558b26c0e6)

With Cython, it is also possible to take advantage of the C++ language, notably,
part of the C++ standard library is directly importable from Cython code.

> Let's see what our code becomes when
> using `vector <https://en.cppreference.com/w/cpp/container/vector>`_ from the C++
> standard library.

Reviewer 3: ```suggestion
Let's see what our code becomes when using
`vector <https://en.cppreference.com/w/cpp/container/vector>`_ from the C++
standard library.
```

Author: Nothing changes from the point of view of the reader of the documentation. If you wanted to make it easier to change and view the source rst, then you also had to wrap the end of the line, i.e. "from the C ++".

Reviewer 4: Yes, I think it helps the maintenance to format it that way.

Author: resolved [here](https://github.com/cython/cython/pull/4226/commits/74e39646d3a0a006c16320ef428d26d0c4fcc147)

    how many elements you are going to put in the vector. For more details
    see `this page from cppreference <https://en.cppreference.com/w/cpp/container/vector>`_.

.. tabs::
    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes_cpp.py
            :linenos:

        .. warning::

            This code uses ``cimports``, cython understands this
            and imports the necessary modules, but python cannot do this,
            therefore it will not work and will raise an exception.

            And since there is no direct analogue of a vector in python,
            with the same functions, we cannot use it if we run it through python.

>             This is described in more detail
>             :ref:`calling-c-functions`.

Reviewer 3: ```suggestion
            This is described in more detail at :ref:`calling-c-functions`.
```

Author: Look [here](https://github.com/cython/cython/issues/4187#issuecomment-861016107). This will be replaced

Author: [replaced](https://github.com/cython/cython/pull/4226/commits/b138847eb1098b57841abe05a6574ca3e827162f)
